unix/x0vncserver/XDesktop.cxx

```diff
@@ -383,6 +391,92 @@ KeyCode XDesktop::XkbKeysymToKeycode(Display* dpy, KeySym keysym) {
 }
 #endif
 
+KeyCode XDesktop::addKeysym(Display* dpy, KeySym keysym)
+{
+  unsigned int key;
+  XkbDescPtr xkb;
+  XkbMapChangesRec changes;
+  KeySym *syms;
+  xkb = XkbGetMap(dpy, XkbAllComponentsMask, XkbUseCoreKbd);
+
+  if (!xkb)
+    return false;
+
+  for (key = xkb->max_key_code; key >= xkb->min_key_code; key--) {
+    if (XkbKeyNumGroups(xkb, key) == 0)
+      break;
+  }
+
+  if (key < xkb->min_key_code)
+    return false;
+
+  memset(&changes, 0, sizeof(changes));
+
+  syms = XkbKeySymsPtr(xkb, key);
+  syms[0] = keysym;
+
+  int types[XkbNumKbdGroups] = { XkbOneLevelIndex };
```

Comment:
Did you check the Xvnc code? It has a comment saying that this doesn't work properly with CapsLock.